polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -1028,18 +1028,26 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				);
 				Ok(())
 			},
-			DescendOrigin(who) => self
-				.context
-				.origin
-				.as_mut()
-				.ok_or(XcmError::BadOrigin)?
-				.append_with(who)
-				.map_err(|e| {
-					tracing::error!(target: "xcm::process_instruction::descend_origin", ?e, "Failed to append junctions");
-					XcmError::LocationFull
-				}),
-			ClearOrigin => {
-				self.context.origin = None;
+			DescendOrigin(who) => self.do_descend_origin(who),
+			ClearOrigin => self.do_clear_origin(),
+			ExecuteWithOrigin { origin, xcm } => {
+				let previous_origin = self.context.origin.clone();
+
+				// Set new temporary origin.
+				if let Some(who) = origin {
+					self.do_descend_origin(who)?;
+				} else {
+					self.do_clear_origin()?;
+				}
+				// Process instructions.
+				self.process(xcm).map_err(|error| {
+					tracing::error!(target: "xcm::execute", ?error, origin = ?self.context.origin, "ExecuteWithOrigin inner xcm failure");
+					self.context.origin = previous_origin.clone();
+					error.xcm_error
+				})?;
+				// Reset origin to previous one.
+				self.context.origin = previous_origin;
```

Comment:
nit: you can avoid clone and resetting origin twice:

```suggestion
				// Process instructions.
				let result = self.process(xcm).map_err(|error| {
					tracing::error!(target: "xcm::execute", ?error, origin = ?self.context.origin, "ExecuteWithOrigin inner xcm failure");
					error.xcm_error
				});
				// Reset origin to previous one.
				self.context.origin = previous_origin;
				result
```